### PR TITLE
CDAP-13254 - Adding Metadata Dataset Migration Service

### PR DESCRIPTION
**Pull request by CuriousVini**
TL;DR:

Implemented MetadataMigrator Service to migrate metadata to V2 tables. In a single transaction, below steps are followed for data migration:
1.) Scan value and history rows from v1 table in batches
2.) Write the rows in new format to v2 table
3.) Delete all the written rows from v1 tables.

Remaining Tasks: 

- Currently testing on the cluster. 
- Handling Queries (Will open separate PR for it).

### CODE REVIEW

File: cdap-api/src/main/java/co/cask/cdap/api/metadata/Metadata.java [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright © 2016 Cask Data, Inc.
+ * Copyright © 2016-2018 Cask Data, Inc.
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/metadata/Metadata.java:2]: unnecessary change?

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataMigrator.java [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+  protected void startUp() throws Exception {
+    LOG.info("Starting Metadata Migrator Service.");
+  }
+
+  @Override
+  public void run() {
+    runThread = Thread.currentThread();
+
+    try {
+      for (int i = 0; i < DATASET_IDS.size(); i = i + 2) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataMigrator.java]: why not use a `Map` instead of `+2` in a `List`?

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataMigrator.java [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+                                                                      NamespaceId.SYSTEM.dataset("business.metadata"),
+                                                                      NamespaceId.SYSTEM.dataset("v2.business.metadata")
+  );
+
+  private final DatasetFramework dsFramework;
+  private final Transactional transactional;
+  private final int batchSize;
+  private volatile Thread runThread;
+  private volatile boolean stopped;
+  private volatile boolean hasV1Instance;
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataMigrator.java]: why this is volatile?

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataMigrator.java [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+
+        while (!stopped && hasV1Instance) {
+          try {
+            final int index = i;
+            // This thread migrates metadata in batches. It does following operations in a single transaction:
+            // 1.) Scan V1 metadata table for value and history rows in batch of batchSize.
+            // 2.) Write scanned MetadataEntries to V2 table.
+            // 3.) Delete successfully written metadata entries from V1 table.
+            Transactionals.execute(transactional, context -> {
+              if (!dsFramework.hasInstance(DATASET_IDS.get(index))) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataMigrator.java]: This check can be outside of the `Transactionals.execute`, and I think it can even be outside of the while loop. In fact, the `hasV1Instance` can be a local variable.

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataService.java [hunk 5]
```diff
@@ -81,6 +91,12 @@ protected void startUp() throws Exception {
 
     httpService.start();
 
+    // Start migration only on the first instance of Dataset Ops Executor.
+    if (instanceId == 0) {
+      MetadataMigrator metadataMigrator = new MetadataMigrator(cConf, dsFramework, txClient);
+      metadataMigrator.start();
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataService.java]: We should stop the metadata migrator when this service is getting stopped.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java [hunk 6]
```diff
@@ -917,4 +931,105 @@ private boolean deleteIndexRow(Row row) {
...
+  /**
+   * Starts scanning/deleting of V1 table value row keys in batches. If there are no more value row keys available,
+   * it will scan history rows. If there are no value/history rows, it returns empty list.
+   *
+   * @param limit Batch size for scan/delete.
+   * @param delete Determines if operation to perform is scan or delete.
+   * @return List of {@link KeyValue} where key is timestamp for history row other wise it is null,
+   * value is {@link MetadataEntry}
+   */
+  public List<KeyValue<Long, MetadataEntry>> scanOrDeleteFromV1Table(int limit, boolean delete) {
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java]: This is an awkward method name (scanOrDelete). I think it's better to have two public methods, one for scanning, one for deleting. They both can be implemented using a shared private method.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java [hunk 7]
```diff
@@ -917,4 +931,105 @@ private boolean deleteIndexRow(Row row) {
...
+      stopRowKey = Bytes.stopKeyForPrefix(rowPrefix);
+      entries.addAll(scanOrDelete(rowPrefix, stopRowKey, limit - entries.size(), delete, this::convertFromV1History));
+    }
+
+    return entries;
+  }
+
+  private List<KeyValue<Long, MetadataEntry>> scanOrDelete(byte[] rowPrefix, byte[] stopRowKey,
+                                                           int limit, boolean delete,
+                                                           Function<Row, KeyValue<Long, MetadataEntry>> function) {
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java]: Since this method is already taking a `Function`. I think it's better call itself `scanAndProcess` and shouldn't explicitly has a `delete` option, but rather the deletion is done via the `function`. Similarly, this method shouldn't return a `List`, but rather having the `function` does the collection if needed.

Comment (chtyim) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java]: And potentially, I think you might want to use `Consumer` instead of `Function`.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataEntry.java [hunk 8]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2015 Cask Data, Inc.
+ * Copyright 2015-2018 Cask Data, Inc.
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataEntry.java:2]: unnecessary change?

**Review by chtyim — COMMENTED** [review]
Did a quick scan. Have some comments. For the correctness of the logic, please have @rohitsinha54 and @anew to review.

File: cdap-api/src/main/java/co/cask/cdap/api/metadata/Metadata.java (hunk 1 above)
Comment (CuriousVini) [reply cdap-api/src/main/java/co/cask/cdap/api/metadata/Metadata.java:2]: There were couple of files changed in 2018 without license year changes.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataEntry.java (hunk 8 above)
Comment (CuriousVini) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataEntry.java:2]: same as above, the file was changed in 2018 but license was not updated.

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataMigrator.java (hunk 3 above)
Comment (CuriousVini) [reply cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataMigrator.java]: yea, there is no need for this to be volatile. As per the comment below, I will make it a local variable

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataMigrator.java (hunk 4 above)
Comment (CuriousVini) [reply cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataMigrator.java]: +1 for calling `hasInstance` outside transaction. But keeping this in the `while` loop will help catch the exception and retry.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java (hunk 7 above)
Comment (CuriousVini) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java]: yea, that seems better implementation, thanks for the suggestion!

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataMigrator.java (hunk 4 above)
Comment (anew) [reply cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataMigrator.java]: Do you assume that if one of the v1 tables doesn't exist, then the other one also does not? I wonder whether it can happen that you process all of system.metadata (v1), and delete it, but while processing business.metadata, a shutdown or failure happens. Then when you come back, you will see that that first v1 table does not exist and return, and never migrate the remaining data?

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataMigrator.java [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+
+              // All the metadata entries are written using setProperty because it does not modify the MetadataEntry
+              // keys
+              List<KeyValue<Long, MetadataEntry>> entries = v1.scanOrDeleteFromV1Table(batchSize, false);
+
+              if (entries.isEmpty()) {
+                // All the value and history rows have been migrated so stop this thread and drop V1 MetadataDataset
+                dropV1MetadataDataset(index);
+                LOG.debug("Migration for dataset {} is complete. This dataset is dropped.", DATASET_IDS.get(index));
+                return;
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataMigrator.java]: so after the first v1 table (system.metadata) is upgraded, it returns? Then the other one will never be upgraded? Maybe you just want to break from this inner loop so that the outer loop can move on to the next dataset?

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataMigrator.java [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+              v1.scanOrDeleteFromV1Table(entries.size(), true);
+            });
+          } catch (Exception e) {
+            OUTAGE_LOG.error("Exception while migrating metadata from {} to {}, will be retried. ", DATASET_IDS.get(i),
+                             DATASET_IDS.get(i + 1), e);
+            Thread.sleep(1000);
+          }
+        }
+      }
+    } catch (InterruptedException e) {
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataMigrator.java]: If you are in the middle of a transaction, will the interrupt be wrapped into a TransactionFailureException? If so, you would not recognize it here.

Comment (anew) [reply cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataMigrator.java]: Actually the catch (Exception e) a few lines up would catch the InterruptedExn, log it, sleep and retry.

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataMigrator.java (hunk 9 above)
Comment (anew) [reply cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataMigrator.java]: oh never mind, this is the return from execute() not the main method

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java [hunk 11]
```diff
@@ -795,10 +802,17 @@ private Put getIndexPut(MetadataEntity metadataEntity, String metadataKey, Strin
    * @param metadataEntity target id for which metadata needs snapshotting
    */
   private void writeHistory(MetadataEntity metadataEntity) {
+    writeHistory(metadataEntity, -1);
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java]: I would find it a little more straight-forward to pass in System.currentTimeMillis(), instead of dealing with the special case of time<0 in the next method

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataMigrator.java [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+              if (entries.isEmpty()) {
+                // All the value and history rows have been migrated so stop this thread and drop V1 MetadataDataset
+                dropV1MetadataDataset(index);
+                LOG.debug("Migration for dataset {} is complete. This dataset is dropped.", DATASET_IDS.get(index));
+                return;
+              }
+
+              v2.writeUpgradedRow(entries);
+              // We do not need to keep checkpoints. Instead, we will just delete scanned rows from metadata dataset
+              v1.scanOrDeleteFromV1Table(entries.size(), true);
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataMigrator.java]: when deleting, do you actually have to scan again? Or can you just do a batch delete of all row keys that you scanned in the first round? That might require remembering the actual keys for each entry, but seems more efficient to me.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java (hunk 6 above)
Comment (anew) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java]: Please see my comment further up. You probably don't need to scan in order to delete, because you already know the row keys from the first scan.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java [hunk 13]
```diff
@@ -795,10 +802,17 @@ private Put getIndexPut(MetadataEntity metadataEntity, String metadataKey, Strin
...
+  }
+
+  private void writeHistory(MetadataEntity metadataEntity, long time) {
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java]: it looks like there is no place in your code, where you actually call this, except for right above. So time is always -1? Then why is this parameter needed?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java [hunk 14]
```diff
@@ -917,4 +931,108 @@ private boolean deleteIndexRow(Row row) {
...
+   * @param entries list of entries to be written.
+   */
+  public void writeUpgradedRow(List<KeyValue<Long, MetadataEntry>> entries) {
+    for (KeyValue<Long, MetadataEntry> kv : entries) {
+      MetadataEntry entry = kv.getValue();
+      if (kv.getKey() == null) {
+        writeWithoutHistory(entry.getMetadataEntity(), entry,
+                            getIndexersForKey(entry.getMetadataEntity(), entry.getKey()));
+      } else {
+        writeHistory(entry.getMetadataEntity());
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java]: i think you want to call ```writeHistory(entry.getMetadataEntity(), kv.getKey()));``` so that the upgraded history entry has the same time as the original one?

**Review by anew — COMMENTED** [review]
please see my comments

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataMigrator.java [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+    runThread = Thread.currentThread();
+
+    try {
+      for (Map.Entry<DatasetId, DatasetId> datasetIdEntry : DATASET_IDS.entrySet()) {
+        // assume we already have v1 metadata table instance
+        AtomicBoolean hasV1Instance = new AtomicBoolean(true);
+
+        while (!stopped && hasV1Instance.get()) {
+          try {
+            if (!dsFramework.hasInstance(datasetIdEntry.getKey())) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataMigrator.java]: Why not just check this outside of the while loop?

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataMigrator.java [hunk 16]
```diff
@@ -0,0 +1,175 @@
...
+  }
+
+  @Override
+  public void run() {
+    runThread = Thread.currentThread();
+
+    try {
+      for (Map.Entry<DatasetId, DatasetId> datasetIdEntry : DATASET_IDS.entrySet()) {
+        // assume we already have v1 metadata table instance
+        AtomicBoolean hasV1Instance = new AtomicBoolean(true);
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataMigrator.java]: Why need to be an atomic boolean? A simple boolean should serve the purpose if you change the `Transactions.execute` to use `TxCallable` to return a boolean.

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataMigrator.java (hunk 4 above)
Comment (CuriousVini) [reply cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataMigrator.java]: The return is from the `Transactional.execute()`. I have taken it out of the transaction (which uses `break`) as per the comment from @chtyim .

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataServiceModule.java [hunk 17]
```diff
@@ -24,16 +24,24 @@
...
 
 import java.util.Set;
 
 /**
  * Guice module for metadata service.
  */
 public class MetadataServiceModule extends PrivateModule {
+  private final Integer instanceId;
+
+  public MetadataServiceModule(TwillContext twillContext) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataServiceModule.java]: If you just need the instanceId, you should just take an instanceId, not the TwillContext. That would keep you better abstraction.

Comment (chtyim) [reply cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataServiceModule.java]: Actually it seems quite strange to inject instance id. I think it's better be just reading a boolean from `cConf` to decide if upgrade logic is needed. Then you can control it via setting the boolean in the `DatasetOpExecutorServerTwillRunnable.createInjector` (it can based on instanceId, or it can based on something else). Also, in standalone, without doing anything, it'll be `false` (assuming you use `cConf.getBoolean("someKey", false);`

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataMigrator.java (hunk 10 above)
Comment (CuriousVini) [reply cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataMigrator.java]: So once the `TransactionFailureException` has thrown, it will be logged and the thread will sleep. In the next retry `stopped` will be true so it wont retry and control will exit the `run()` method.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java (hunk 13 above)
Comment (CuriousVini) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java]: Oh good catch, it should be called in `writeUpgradedRow` because we want to use same timestamp for history rows.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java (hunk 14 above)
Comment (CuriousVini) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java]: yea, my bad!

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataMigrator.java (hunk 12 above)
Comment (CuriousVini) [reply cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataMigrator.java]: yea not scanning again would be efficient.

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataMigrator.java (hunk 10 above)
Comment (anew) [reply cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataMigrator.java]: ok

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MdsHistoryKey.java [hunk 18]
```diff
@@ -36,6 +40,49 @@ public static MDSKey getMdsKey(NamespacedEntityId targetId, long time) {
...
+   * @param type type of the entity
+   * @return timestamp in the rowKey
+   */
+  static long extractTime(byte[] rowKey, String type) {
+    MDSKey.Splitter keySplitter = new MDSKey(rowKey).split();
+    // Skip rowType
+    keySplitter.skipBytes();
+
+    // Skip targetId
+    if (type.equals(EntityIdKeyHelper.TYPE_MAP.get(ProgramId.class))) {
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MdsHistoryKey.java]: If I understand correctly, then ```EntityIdKeyHelper.TYPE_MAP.get()``` returns a String representing a EntityTypeSimpleName. So you could write this as a 

```
switch(EntityTypeSimpleName.valueOf(type)) {
case: PROGRAM: 
  // call skipString 4 times; 
  break;
case APPLICATION:
case DATASET:
case STREAM:
  // skipString() 2x
  break;
...
}
```

Wouldn't that be simpler?

File: cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataMigratorTest.java [hunk 19]
```diff
@@ -153,7 +154,7 @@ public void testMetadataMigration() throws Exception {
     MetadataMigrator migrator = new MetadataMigrator(cConf, datasetFramework, transactionSystemClient);
     migrator.start();
 
-    // Wait for migrator to finish before reading v2 tables
+    // Wait for migrator to finish before reading v2 tablesf
```
Comment (CuriousVini) [inline cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataMigratorTest.java]: remove this unnecessary change in latest commit

**Comment by CuriousVini** [comment]
Build: https://builds.cask.co/browse/CDAP-DUT6514-6

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MdsHistoryKey.java [hunk 20]
```diff
@@ -51,33 +46,37 @@ static long extractTime(byte[] rowKey, String type) {
...
-      keySplitter.skipString();
-    } else if (type.equals(EntityIdKeyHelper.TYPE_MAP.get(ArtifactId.class))) {
-      // skip namespace, name, version
-      keySplitter.skipString();
-      keySplitter.skipString();
-      keySplitter.skipString();
-    } else {
-      throw new IllegalArgumentException("Illegal Type " + type + " of metadata source.");
+    switch (type) {
+      case MetadataEntity.PROGRAM:
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MdsHistoryKey.java]: you can further simplify this by merging all cases that skip two strings into one:

```
case APPLICATION:
case DATASET:
case STREAM:
    skipString();
    skipString();
    break;
case VIEW:
case ARTIFACT:
    skipString();
    skipString();
    skipString();
    break;
case PROGRAM:
    skipString();
    skipString();
    skipString();
    skipString();
    break;
default: ....
```

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MdsHistoryKey.java [hunk 21]
```diff
@@ -47,30 +53,30 @@ static long extractTime(byte[] rowKey, String type) {
...
     switch (type) {
-      case MetadataEntity.PROGRAM:
+      case PROGRAM:
         keySplitter.skipString();
         keySplitter.skipString();
         keySplitter.skipString();
         keySplitter.skipString();
         break;
-      case MetadataEntity.APPLICATION:
+      case APPLICATION:
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MdsHistoryKey.java]: you can still avoid repetition by using multiple case labels. Please see my original comment.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MdsHistoryKey.java [hunk 22]
```diff
@@ -16,11 +16,17 @@
...
 package co.cask.cdap.data2.metadata.dataset;
 
-import co.cask.cdap.api.metadata.MetadataEntity;
 import co.cask.cdap.data2.dataset2.lib.table.EntityIdKeyHelper;
 import co.cask.cdap.data2.dataset2.lib.table.MDSKey;
 import co.cask.cdap.proto.id.NamespacedEntityId;
 
+import static co.cask.cdap.api.metadata.MetadataEntity.APPLICATION;
+import static co.cask.cdap.api.metadata.MetadataEntity.ARTIFACT;
+import static co.cask.cdap.api.metadata.MetadataEntity.DATASET;
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MdsHistoryKey.java]: does this also need to account for DatsetInstance?

Comment (anew) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MdsHistoryKey.java]: oh I see that was done in second commit

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MdsKey.java [hunk 23]
```diff
@@ -57,7 +58,8 @@ static String getMetadataKey(String type, byte[] rowKey) {
     } else if (type.equals(EntityIdKeyHelper.TYPE_MAP.get(ApplicationId.class))) {
       keySplitter.skipString();
       keySplitter.skipString();
-    } else if (type.equals(EntityIdKeyHelper.TYPE_MAP.get(DatasetId.class))) {
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MdsKey.java]: can't this method be simplified in the same way as the one above in MdsHistoryKey?

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataMigrator.java [hunk 24]
```diff
@@ -0,0 +1,152 @@
...
+ * Metadata Migrator to migrate metadata from V1 metadata tables to V2 metadata tables.
+ * Migration happens on business table as well as system table. It only migrates value and history rows.
+ * Indexes will be automatically generated for every write on V2 metadata table.
+ */
+class MetadataMigrator extends AbstractRetryableScheduledService {
+  private static final Logger LOG = LoggerFactory.getLogger(MetadataMigrator.class);
+  // For outage, only log once per 60 seconds per message.
+  private static final Logger OUTAGE_LOG = Loggers.sampling(LOG, LogSamplers.perMessage(
+    () -> LogSamplers.limitRate(60000)));
+  private static final Queue<KeyValue<DatasetId, DatasetId>> DATASETS = new LinkedList<>();
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataMigrator.java]: Since you are modifying the queue in the instance, it is better to have this variable as a field, rather than static field.

**Review by chtyim — COMMENTED** [review]
Just one minor comment, the rest LGTM

**Comment by CuriousVini** [comment]
Waiting for the build: https://builds.cask.co/browse/CDAP-RUT1510-6

**Comment by CuriousVini** [comment]
The build is green. Merging.
